### PR TITLE
feature: timeout on request

### PR DESCRIPTION
**Pull request by FrankEssenberger**
relates SAP/cloud-sdk-backlog#83
relates SAP/cloud-sdk-backlog#79

Creates two timesouts:
- For the requests against the BTP services xsuaa and destination-service. This timeout is controlled via the destinationFetchOptions
- For the requests against the target system given in the destination. This timeout is controlled either by the request builder or the request config for the executeHttpRequest.

### CODE REVIEW

File: packages/connectivity/src/scp-cf/token-accessor.spec.ts [hunk 1]
```diff
@@ -47,6 +49,51 @@ describe('token accessor', () => {
       expect(actual).toBe(expected);
     });
 
+    it('considers custom timeout for client credentials token', async () => {
+      const jwt = signedJwt({
+        iss: 'https://testeroni.example.com'
+      });
+      async function doDelayTest(enableCircuitBreaker: boolean) {
```
Comment (florian-richter) [inline packages/connectivity/src/scp-cf/token-accessor.spec.ts:56]:
```suggestion
      async function doDelayTest() {
```

File: packages/connectivity/src/scp-cf/token-accessor.spec.ts [hunk 2]
```diff
@@ -47,6 +49,51 @@ describe('token accessor', () => {
...
+        iss: 'https://testeroni.example.com'
+      });
+      mockClientCredentialsGrantCall(
+        'https://testeroni.example.com',
+        { access_token: 'testValue' },
+        200,
+        destinationBindingClientSecretMock.credentials
+      );
+
+      const token = await serviceToken('destination', { jwt });
```
Comment (florian-richter) [inline packages/connectivity/src/scp-cf/token-accessor.spec.ts:90]:
```suggestion
      await serviceToken('destination', { jwt });
```

File: packages/http-client/src/http-client.ts [hunk 3]
```diff
@@ -445,6 +451,7 @@ async function getCsrfHeaders(
         params: request.params,
         headers: request.headers,
         url: request.url,
+        timeout: request.timeout,
```
Comment (florian-richter) [inline packages/http-client/src/http-client.ts:454]:
```suggestion
        timeout: request.timeout || defaultTimeoutTarget,
```

The last commit (which changed this) seems to have broken a test.

File: packages/connectivity/src/scp-cf/token-accessor.spec.ts (hunk 1 above)
Comment (deekshas8) [reply packages/connectivity/src/scp-cf/token-accessor.spec.ts:56]: I think `enableCircuitBreaker` has to be passed as an option to `serviceToken` below.

File: packages/odata-common/src/request-builder/request-builder-base.ts [hunk 4]
```diff
@@ -48,6 +48,11 @@ export abstract class MethodRequestBuilder<
     return this;
   }
 
+  timeout(timeout: number): this {
```
Comment (deekshas8) [inline packages/odata-common/src/request-builder/request-builder-base.ts:51]: Please consider adding a doc comment and also using a verb in the function name.

File: packages/connectivity/src/scp-cf/xsuaa-service.ts [hunk 5]
```diff
@@ -90,27 +91,38 @@ export function getSubdomainAndZoneId(
...
-    xssec.requests.requestClientCredentialsToken(
-      subdomainAndZoneId.subdomain,
-      serviceCredentials,
-      null,
-      subdomainAndZoneId.zoneId,
-      (err: Error, token: string, tokenResponse: ClientCredentialsResponse) =>
-        err ? reject(err) : resolve(tokenResponse)
-    );
-  });
+  const xssePromise: Promise<ClientCredentialsResponse> = new Promise(
```
Comment (deekshas8) [inline packages/connectivity/src/scp-cf/xsuaa-service.ts:112]:
```suggestion
  const xssecPromise: Promise<ClientCredentialsResponse> = new Promise(
```

File: packages/connectivity/src/scp-cf/xsuaa-service.ts [hunk 6]
```diff
@@ -123,14 +135,21 @@ export function getClientCredentialsToken(
...
+  if (enableCircuitBreaker) {
+    return wrapInCircuitBreaker(getUserToken)(service, userJwt, {
+      enableCircuitBreaker: false,
+      timeout
+    });
   }
   const subdomainAndZoneId = getSubdomainAndZoneId(userJwt);
 
-  return new Promise((resolve: (token: string) => void, reject) =>
+  const xssePromise = new Promise((resolve: (token: string) => void, reject) =>
```
Comment (deekshas8) [inline packages/connectivity/src/scp-cf/xsuaa-service.ts:152]:
```suggestion
  const xssecPromise = new Promise((resolve: (token: string) => void, reject) =>
```

File: packages/connectivity/src/scp-cf/token-accessor.spec.ts (hunk 1 above)
Comment (deekshas8) [reply packages/connectivity/src/scp-cf/token-accessor.spec.ts:56]: I think `enableCircuitBreaker` has to be passed as an option in the call to `serviceToken`.

**Review by deekshas8 — APPROVED** [review]
LGTM.

**Pull request merged**